### ISSUE TITLE
[AMI 0.18] Longevity test failure: Failed to decommission node (scylla coredump generated)

### ISSUE THREAD

**Issue by lmr** [scylladb/scylla#924]
When running the 24 hour longevity test with the decommission node (more on that later), after a few successful decomission attempts, we start hitting failures.

The first couple of hours things go well, and decommission happens flawlessly, taking ~ 650 s to complete:

    17:13:09 nemesis          L0190 DEBUG| sdcm.nemesis.DecommissionMonkey: <function disrupt at 0x7f19890a0a28> duration -> 683 s
    17:53:56 nemesis          L0190 DEBUG| sdcm.nemesis.DecommissionMonkey: <function disrupt at 0x7f19890a0a28> duration -> 647 s
    18:34:00 nemesis          L0190 DEBUG| sdcm.nemesis.DecommissionMonkey: <function disrupt at 0x7f19890a0a28> duration -> 603 s
    19:16:13 nemesis          L0190 DEBUG| sdcm.nemesis.DecommissionMonkey: <function disrupt at 0x7f19890a0a28> duration -> 733 s
    19:58:17 nemesis          L0190 DEBUG| sdcm.nemesis.DecommissionMonkey: <function disrupt at 0x7f19890a0a28> duration -> 723 s

Then things go south:

    20:29:24 nemesis          L0069 ERROR| sdcm.nemesis.DecommissionMonkey: nodetool command 'nodetool --host localhost decommission' failed on node Node decommission_monkey-scylla-db-node-52fdc9f4-6 [54.187.230.144 | 172.31.39.48] (seed: False): Command: /bin/ssh -a -x  -o ControlPath=/var/tmp/ssh-masteroO1egv/socket -o StrictHostKeyChecking=no -o UserKnownHostsFile=/var/tmp/tmp7Cmh_k -o BatchMode=yes -o ConnectTimeout=300 -o ServerAliveInterval=300 -l centos -p 22 -i /var/tmp/decommission_monkey-longevity-test-0a47d94a.pem 54.187.230.144 "nodetool --host localhost decommission"
    Exit status: 2
    Duration: 66.2351438999
    Stdout:

    Stderr:
    error: javax.ws.rs.ProcessingException (no security manager: RMI class loader disabled)
    -- StackTrace --
    java.lang.ClassNotFoundException: javax.ws.rs.ProcessingException (no security manager: RMI class loader disabled)
            at sun.rmi.server.LoaderHandler.loadClass(LoaderHandler.java:396)
            at sun.rmi.server.LoaderHandler.loadClass(LoaderHandler.java:186)
            at java.rmi.server.RMIClassLoader$2.loadClass(RMIClassLoader.java:637)
            at java.rmi.server.RMIClassLoader.loadClass(RMIClassLoader.java:264)
            at sun.rmi.server.MarshalInputStream.resolveClass(MarshalInputStream.java:214)
            at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1613)
            at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1518)
            at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1774)
            at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
            at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2000)
            at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:1924)
            at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1801)
            at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1351)
            at java.io.ObjectInputStream.readObject(ObjectInputStream.java:371)
            at sun.rmi.transport.StreamRemoteCall.executeCall(StreamRemoteCall.java:245)
            at sun.rmi.server.UnicastRef.invoke(UnicastRef.java:162)
            at com.sun.jmx.remote.internal.PRef.invoke(Unknown Source)
            at javax.management.remote.rmi.RMIConnectionImpl_Stub.invoke(Unknown Source)
            at javax.management.remote.rmi.RMIConnector$RemoteMBeanServerConnection.invoke(RMIConnector.java:1020)
            at javax.management.MBeanServerInvocationHandler.invoke(MBeanServerInvocationHandler.java:298)
            at com.sun.proxy.$Proxy8.decommission(Unknown Source)
            at org.apache.cassandra.tools.NodeProbe.decommission(NodeProbe.java:606)
            at org.apache.cassandra.tools.NodeTool$Decommission.execute(NodeTool.java:1460)
            at org.apache.cassandra.tools.NodeTool$NodeToolCmd.run(NodeTool.java:288)
            at org.apache.cassandra.tools.NodeTool.main(NodeTool.java:202)

    20:29:24 nemesis          L0190 DEBUG| sdcm.nemesis.DecommissionMonkey: <function disrupt at 0x7f19890a0a28> duration -> 67 s

Then in the following occasions, the JMX server stops responding altogether:

    20:59:27 nemesis          L0069 ERROR| sdcm.nemesis.DecommissionMonkey: nodetool command 'nodetool --host localhost decommission' failed on node Node decommission_monkey-scylla-db-node-52fdc9f4-6 [54.187.230.144 | 172.31.39.48] (seed: False): Command: /bin/ssh -a -x  -o ControlPath=/var/tmp/ssh-masteroO1egv/socket -o StrictHostKeyChecking=no -o UserKnownHostsFile=/var/tmp/tmp7Cmh_k -o BatchMode=yes -o ConnectTimeout=300 -o ServerAliveInterval=300 -l centos -p 22 -i /var/tmp/decommission_monkey-longevity-test-0a47d94a.pem 54.187.230.144 "nodetool --host localhost decommission"
    Exit status: 1
    Duration: 2.21429300308
    Stdout:

    Stderr:
    nodetool: Failed to connect to 'localhost:7199' - ConnectException: 'Connection refused'.

Analyzing logs, we discovered that the failure to execute decommission on the node was due to scylla-server.service dumping core about the time the nodetool command was invoked:

               PID: 2799 (scylla)
               UID: 994 (scylla)
               GID: 1001 (scylla)
            Signal: 11 (SEGV)
         Timestamp: Fri 2016-02-19 20:28:19 UTC (7min ago)
      Command Line: /usr/bin/scylla
        Executable: /usr/bin/scylla
     Control Group: /system.slice/scylla-server.service
              Unit: scylla-server.service
             Slice: system.slice
           Boot ID: af18573c919b44feb4eaaf627721f8af
        Machine ID: 95e7cd3c64fe41179400cdc37d961c5e
          Hostname: ip-172-31-39-48
          Coredump: /var/lib/systemd/coredump/core.scylla.994.af18573c919b44feb4eaaf627721f8af.2799.1455913699000000.xz
           Message: Process 2799 (scylla) of user 994 dumped core.
                    
                    Stack trace of thread 2799:
                    #0  0x00000000007f9b83 _ZN13managed_bytesC2ERKS_ (scylla)
                    #1  0x0000000000d92fbf _ZN3dht5tokenC4ERKS0_ (scylla)
                    #2  0x0000000000ffbe63 _ZN7service32pending_range_calculator_service24calculate_pending_rangesERN7locator29abstract_replication_strategyERK13basic_sstringIcjLj15EE (scylla)
                    #3  0x0000000000ffc032 operator() (scylla)
                    #4  0x0000000000efb494 operator() (scylla)
                    #5  0x00000000004c9f4d _ZN7reactor9run_tasksER15circular_bufferISt10unique_ptrI4taskSt14default_deleteIS2_EESaIS5_EE (scylla)
                    #6  0x00000000004f477b _ZN7reactor3runEv (scylla)
                    #7  0x000000000056b188 _ZN12app_template14run_deprecatedEiPPcOSt8functionIFvvEE (scylla)
                    #8  0x000000000047043b main (scylla)
                    #9  0x00007f17d539ab15 __libc_start_main (libc.so.6)
                    #10 0x00000000004c6571 _start (scylla)
                    
                    Stack trace of thread 2806:
                    #0  0x00007f17d574822d read (libpthread.so.0)
                    #1  0x00000000004ca695 _ZN11thread_pool4workEv (scylla)
                    #2  0x00000000005569de _ZNKSt8functionIFvvEEclEv (scylla)
                    #3  0x00007f17d5741dc5 start_thread (libpthread.so.0)
                    #4  0x00007f17d546f28d __clone (libc.so.6)
                    
                    Stack trace of thread 2803:
                    #0  0x00007f17d546f863 epoll_wait (libc.so.6)
                    #1  0x00000000006c6014 eal_intr_thread_main (scylla)
                    #2  0x00007f17d5741dc5 start_thread (libpthread.so.0)
                    #3  0x00007f17d546f28d __clone (libc.so.6)
                    
                    Stack trace of thread 2804:
                    #0  0x00007f17d546f4c7 epoll_pwait (libc.so.6)
                    #1  0x00000000004d4b3b _ZN21reactor_backend_epoll16wait_and_processEiPK10__sigset_t (scylla)
                    #2  0x00000000004d4cb1 _ZN7reactor16wait_and_processEiPK10__sigset_t (scylla)
                    #3  0x00000000004f4b13 _ZN7reactor3runEv (scylla)
                    #4  0x0000000000516839 _ZZN3smp9configureEN5boost15program_options13variables_mapEENKUlvE1_clEv.constprop.3316 (scylla)
                    #5  0x00000000004c6cfe _ZNKSt8functionIFvvEEclEv (scylla)
                    #6  0x00000000006c28b5 eal_thread_loop (scylla)
                    #7  0x00007f17d5741dc5 start_thread (libpthread.so.0)
                    #8  0x00007f17d546f28d __clone (libc.so.6)
                    
                    Stack trace of thread 2805:
                    #0  0x00007f17d574822d read (libpthread.so.0)
                    #1  0x00000000004ca695 _ZN11thread_pool4workEv (scylla)
                    #2  0x00000000005569de _ZNKSt8functionIFvvEEclEv (scylla)
                    #3  0x00007f17d5741dc5 start_thread (libpthread.so.0)
                    #4  0x00007f17d546f28d __clone (libc.so.6)

Then that node runs out of space due to the massive core dump generated.

What the test does
------------------

The workflow of a longevity test is:

1. Create a scylla cluster (referred as DB cluster from now on) with number of nodes (default size is 6) and node size (default is t2.micro, the bug report is using c3.large) defined by the user on a config file
2. Create a number of loader nodes, that will stress the DB cluster, typically by running cassandra stress on it
3. The cluster object hosts a Nemesis (that terminology was borrowed from Jepsen), a thread that randomly picks a non-seed node and disrupt it in some way.
4. Nemesis will sleep for a specified time interval, then execute the disruption function, then sleep again, rinse and repeat.
5. Cassandra-stress starts. The default command line looks like:

    `cassandra-stress write cl=QUORUM duration=1440m -schema 'replication(factor=3)' -port jmx=6868 -mode cql3 native -rate threads=4 -node 172.31.27.192`

6. At all times, we have at least initial cluster size - 1 (6 -1 = 5) nodes active, which should be enough to keep c-s going. The disruption functions operate on a single node, and operations that actually remove a node permanently from the cluster will add a new node to compensate the one that was removed.
7. Given all that, we do expect that c-s survives until the specified time period (the report is using 24 hours). If it doesn't, chances are that we have a bug.

What the nemesis does
---------------------

1. Execute the following command in the target node:

    `nodetool --host localhost decommission`

2. Verify that the node was removed from the cluster
3. If it was, terminate target node and add a new cluster node to compensate for the node removed

**Comment by lmr** [comment]
By the way, node size is `c3.large`.

**Comment by avikivity** [comment]
Please  upload the core somewhere.

**Comment by lmr** [comment]
@avikivity That's not possible, as the test already ended and cleaned up AWS resources.

We discussed a lot about this, and decided that we'd stick with the traces produced by systemd.

**Comment by avikivity** [comment]
Let's discuss it some more and decide to recover core dumps.

**Comment by avikivity** [comment]
Uploading to S3 is a one-liner.  What's important is to be able to associate a core with a particular test.

Let's use a separate bucket for test cores (vs. user uploaded cores).

**Comment by slivne** [comment]
this seems similar to #911 - that can be locally reproduced with dtest - so lets fix #911 and see if it resolves this as well

**Comment by slivne** [comment]
#911 was fixed need to validate if this is fixed as well - if not @asias will take a look

**Comment by lmr** [comment]
I have repeated this with a smaller nemesis interval (5 minutes), and I could not reproduce the core dump generation. Cassandra-stress doesn't survive, though. I found a new message in the logs that I don't recall seeing before:

    08:02:05 remote           L0182 DEBUG| [54.187.237.202] [stdout] Feb 24 08:02:08 ip-172-31-46-134 scylla[2073]:  [shard 1] gossip - unknown endpoint 172.31.38.235
    08:02:05 remote           L0182 DEBUG| [54.187.237.202] [stdout] Feb 24 08:02:08 ip-172-31-46-134 scylla[2073]:  [shard 1] gossip - unknown endpoint 172.31.38.235
    08:02:05 remote           L0182 DEBUG| [54.187.237.202] [stdout] Feb 24 08:02:08 ip-172-31-46-134 scylla[2073]:  [shard 1] gossip - unknown endpoint 172.31.38.235

**Comment by lmr** [comment]
About the error message, this 'unknown endpoint' is cluster node 6:

    08:00:05 cluster          L0156 DEBUG| Node decommission_monkey-scylla-db-node-4ab2fb3c-6 [54.201.121.122 | 172.31.38.235] (seed: None): SSH access -> 'ssh -i /var/tmp/decommission_monkey-longevity-test-dc3c5817.pem centos@54.201.121.122'

**Comment by asias** [comment]
@lmr since you reduced nemesis interval to 5 minutes, when the next decommission starts, does the previous decommission finish? The log comes from 
```
  0 bool gossiper::is_alive(inet_address ep) {
  1     if (ep == get_broadcast_address()) {
  2         return true;
  3     }
  4     auto eps = get_endpoint_state_for_endpoint(ep);
  5     // we could assert not-null, but having isAlive fail screws a node over so badly that
  6     // it's worth being defensive here so minor bugs don't cause disproportionate
  7     // badness.  (See CASSANDRA-1463 for an example).
  8     if (eps) {
  9         return eps->is_alive();
 10     } else {
 11         logger.warn("unknown endpoint {}", ep);
 12         return false;
 13     }
 14 }
```
The log shows, we called gossip::is_alive against a node which is removed inside gossip.
Is the ip 172.31.38.235 from the node we just decommissioned?

Do you still see the warning if you increase interval a bit, like 15 minutes.

**Comment by lmr** [comment]
@asias, I'm sorry for the late reply, I wanted to retry the tests more times in order to get better answers:

1) The previous decommission finishes. In fact, the interval happens when all the decommissioning operations are complete. The test gives [total test timeout, in this case 650000 s] time for nemesis to work.
2)  Yes, 172.31.38.235 is the node just decommissioned
3) I tried it again with a slightly newer scylla RPM and I no longer see this message
4) The bad news is that now I'm seeing a nodetool decomission that never returns (it's stuck trying to execute the decommission command for hours now). No coredump or command failure though. I'll open another bug.

**Comment by asias** [comment]
On Mar 1, 2016 2:32 AM, "Lucas Meneghel Rodrigues" <notifications@github.com>
wrote:
>
> @asias, I'm sorry for the late reply, I wanted to retry the tests more
times in order to get better answers:
>
> 1) The previous decommission finishes. In fact, the interval happens when
all the decommissioning operations are complete. The test gives [total test
timeout, in this case 650000 s] time for nemesis to work.
> 2) Yes, 172.31.38.235 is the node just decommissioned
> 3) I tried it again with a slightly newer scylla RPM and I no longer see
this message
> 4) The bad news is that now I'm seeing a nodetool decomission that never
returns (it's stuck trying to execute the decommission command for hours
now). No coredump or command failure though. I'll open another bug.

The hang is probably due to the cql shutdown issue. It is a known issue.

**Comment by asias** [comment]
@lmr I think the "unknown endpoint" issue is due to https://github.com/scylladb/scylla/issues/909#issuecomment-185607109 which is now fixed.

decommission hang is due to cql server shutdown hang which should be fixed by #909 and #965. 

The core dump issue from the original report is fixed already. 

I suggest to rerun the test to verify and close this bug.

**Comment by lmr** [comment]
This problem doesn't reproduce anymore as of the latest 1.0-rc1 AMI (ami-c25e51a8). I'm closing this issue.
